File: 30. First Completely Painted Row or Column.cpp
class Solution {
public:
    int firstCompleteIndex(vector<int>& arr, vector<vector<int>>& mat) {
        int m = mat.size();
        int n = mat[0].size();

        // Maps to store the positions of each number in the matrix
        unordered_map<int, pair<int, int>> pos;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                pos[mat[i][j]] = {i, j};
            }
        }

        // Row and column paint counters
        vector<int> rowPaint(m, 0);
        vector<int> colPaint(n, 0);

        // Process arr and track the painting
        for (int i = 0; i < arr.size(); ++i) {
            int num = arr[i];
            auto [r, c] = pos[num];

            // Increment the paint counters for the row and column
            rowPaint[r]++;
            colPaint[c]++;

            // Check if a row or column is fully painted
            if (rowPaint[r] == n || colPaint[c] == m) {
                return i;
            }
        }

        return -1; // This should never be reached
    }
};

File: 30. First Completely Painted Row or Column.py
class Solution(object):
    def firstCompleteIndex(self, arr, mat):
        """
        :type arr: List[int]
        :type mat: List[List[int]]
        :rtype: int
        """
        m, n = len(mat), len(mat[0])
        position = {}
        
        # Store positions of elements in mat
        for i in range(m):
            for j in range(n):
                position[mat[i][j]] = (i, j)
        
        row_count = [0] * m  # Tracks painted cells in each row
        col_count = [0] * n  # Tracks painted cells in each column
        
        # Iterate through arr and paint the respective cell
        for i, num in enumerate(arr):
            r, c = position[num]
            row_count[r] += 1
            col_count[c] += 1
            
            # Check if a row or column is completely painted
            if row_count[r] == n or col_count[c] == m:
                return i
        
        return -1  # This case should never occur as a full row or column will always be painted.
